From 93ecb4a124e971455f816c027095656f342b0488 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Sun, 19 Jan 2025 20:52:09 +0100
Subject: [PATCH 06/11] arm64: dts: ti: k3-am642-hummingboard-t: fix sdcard
 detection timeout

Specify SD-Card vmmc-supply to avoid sdhci-am654 driver enabling
quirk SDHCI_AM654_QUIRK_SUPPRESS_V1P8_ENA.

This quirk is activated when either both or neither vmmc-supply and
vqmmc-supply are defined, causing sd-card detection to time out:

mmc1: error -110 whilst initialising SD card

Fixes: 5bc0054f6e8f ("mmc: sdhci_am654: Add sdhci_am654_start_signal_voltage_switch")
Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 .../boot/dts/ti/k3-am642-hummingboard-t.dts   | 21 +++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/arch/arm64/boot/dts/ti/k3-am642-hummingboard-t.dts b/arch/arm64/boot/dts/ti/k3-am642-hummingboard-t.dts
index 5b5e9eeec5ac..614879bacc2f 100644
--- a/arch/arm64/boot/dts/ti/k3-am642-hummingboard-t.dts
+++ b/arch/arm64/boot/dts/ti/k3-am642-hummingboard-t.dts
@@ -62,6 +62,15 @@ regulator-m2-3v3 {
 		regulator-always-on;
 	};
 
+	vdd_mmc1: regulator-vdd-mmc1 {
+		compatible = "regulator-fixed";
+		regulator-name = "vdd-mmc1";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
+		regulator-boot-on;
+	};
+
 	regulator-vpp-1v8 {
 		compatible = "regulator-fixed";
 		pinctrl-names = "default";
@@ -286,6 +295,18 @@ &main_uart3 {
 	status = "okay";
 };
 
+&sdhci1 {
+	/*
+	 * Specify fixed 3.3V regulator as vmmc-supply to avoid
+	 * SDHCI_AM654_QUIRK_SUPPRESS_V1P8_ENA enabled by sdhci-am654 driver
+	 * when either both or neither vmmc-supply and vqmmc-supply
+	 * are specified.
+	 * With the quirk enabled, sdcard detection times out:
+	 * mmc1: error -110 whilst initialising SD card
+	 */
+	vmmc-supply = <&vdd_mmc1>;
+};
+
 &usb0 {
 	dr_mode = "host";
 };
-- 
2.43.0

